envoy/cilium_network_policy.h

```diff
@@ -0,0 +1,218 @@
+    };
+    
+    class PortNetworkPolicyRule : public Logger::Loggable<Logger::Id::config> {
+    public:
+      PortNetworkPolicyRule(const cilium::PortNetworkPolicyRule& rules) {
+	for (const auto& remote: rules.remote_policies()) {
+	  ENVOY_LOG(trace, "Cilium L7 PortNetworkPolicyRule(): Allowing remote {}", remote);
+	  allowed_remotes_.emplace(remote);
+	}
+	for (const auto& http_rule: rules.http_rules().http_rules()) {
+	  http_rules_.emplace_back(http_rule);
+	}
+      }
+
+      bool Matches(uint64_t remote_id, const Envoy::Http::HeaderMap& headers) const {
+	// Remote ID must match if we have any.
+	if (allowed_remotes_.size()) {
+	  bool matches = false;
+	  for (const auto& remote: allowed_remotes_) {
+	    if (remote == remote_id) {
+	      matches = true;
+	      break;
+	    }
+	  }
+	  if (!matches) {
+	    return false;
+	  }
+	}
+	// Empty set matches any payload
+	if (http_rules_.size() == 0) {
+	  return true;
+	}	
+	for (const auto& rule: http_rules_) {
+	  if (rule.Matches(headers)) {
+	    return true;
+	  }
+	}
+	return false;
+      }
+
+      std::unordered_set<uint64_t> allowed_remotes_; // Everyone allowed if empty.
+      std::vector<HttpNetworkPolicyRule> http_rules_; // Allowed if empty, but remote is checked first.
+    };
+
+    class PortNetworkPolicyRules : public Logger::Loggable<Logger::Id::config> {
+    public:
+      PortNetworkPolicyRules(const google::protobuf::RepeatedPtrField<cilium::PortNetworkPolicyRule>& rules) {
+	if (rules.size() == 0) {
+	    ENVOY_LOG(trace, "Cilium L7 PortNetworkPolicyRules(): No rules, will allow everything.");
+	}
+	for (const auto& it: rules) {
+	  rules_.emplace_back(PortNetworkPolicyRule(it));
+	}
+      }
+
+      bool Matches(uint64_t remote_id, const Envoy::Http::HeaderMap& headers) const {
+	// Empty set matches any payload from anyone
+	if (rules_.size() == 0) {
+	  return true;
+	}
+	for (const auto& rule: rules_) {
+	  if (rule.Matches(remote_id, headers)) {
+	    return true;
+	  }
+	}
+	return false;
+      }
+
+      std::vector<PortNetworkPolicyRule> rules_; // Allowed if empty.
+    };
+    
+    class PortNetworkPolicy : public Logger::Loggable<Logger::Id::config> {
+    public:
+      PortNetworkPolicy(const google::protobuf::RepeatedPtrField<cilium::PortNetworkPolicy>& rules) {
+	for (const auto& it: rules) {
+	  // TODO: Handle Missing Protocol
+	  if (it.protocol() == envoy::api::v2::core::SocketAddress::TCP) {
+	    // port may be zero, which matches any port.
+	    ENVOY_LOG(trace, "Cilium L7 PortNetworkPolicy(): installing TCP policy for port {}", it.port());
+	    rules_.emplace(it.port(), PortNetworkPolicyRules(it.rules()));
+	  } else {
+	    ENVOY_LOG(warn, "Cilium L7 PortNetworkPolicy(): NOT installing non-TCP policy");
+	  }
+	}
+      }
+
+      bool Matches(uint32_t port, uint64_t remote_id, const Envoy::Http::HeaderMap& headers) const {
+	auto range = rules_.equal_range(port);
```

Comment:
I don't understand why you use `equal_range` instead of `find`. There can be at most one entry in the map with that port number, right?

Reply:
Remnant from the time I was planning to support duplicates, will change to a simple `find()`.